Modules/posixmodule.c

```diff
@@ -7651,6 +7651,115 @@ os_sched_getaffinity_impl(PyObject *module, pid_t pid)
 
 #endif /* HAVE_SCHED_H */
 
+#ifdef HAVE_POSIX_OPENPT
+/*[clinic input]
+os.posix_openpt -> int
+
+    oflag: int
+    /
+
+Open and return a file descriptor for a master pseudo-terminal device.
+
+Performs a posix_openpt() system call. The oflag argument is used to set
+file status flags and file access modes as specified in the manual page
+of posix_openpt() of your system.
+[clinic start generated code]*/
+
+static int
+os_posix_openpt_impl(PyObject *module, int oflag)
+/*[clinic end generated code: output=ee0bc2624305fc79 input=012468535878d78e]*/
+{
+    int fd;
+
+    fd = posix_openpt(oflag);
+    if (fd == -1)
+        posix_error();
+
+    return fd;
+}
+#endif /* HAVE_POSIX_OPENPT */
+
+#ifdef HAVE_GRANTPT
+/*[clinic input]
+os.grantpt
+
+    fd: fildes
+        File descriptor of a master pseudo-terminal device.
+    /
+
+Grant access to the slave pseudo-terminal device.
+
+Performs a grantpt() system call.
+[clinic start generated code]*/
+
+static PyObject *
+os_grantpt_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=dfd580015cf548ab input=7516c1a3cc2b9cfc]*/
+{
+    int ret;
+    PyOS_sighandler_t sig_saved;
+
+    sig_saved = PyOS_setsig(SIGCHLD, SIG_DFL);
+    ret = grantpt(fd);
+    PyOS_setsig(SIGCHLD, sig_saved);
+
+    if (ret == -1)
+        return posix_error();
+
+    Py_RETURN_NONE;
+}
+#endif /* HAVE_GRANTPT */
+
+#ifdef HAVE_UNLOCKPT
+/*[clinic input]
+os.unlockpt
+
+    fd: fildes
+        File descriptor of a master pseudo-terminal device.
+    /
+
+Unlock a pseudo-terminal master/slave pair.
+
+Performs an unlockpt() system call.
+[clinic start generated code]*/
+
+static PyObject *
+os_unlockpt_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=e08d354dec12d30c input=125f704b2c0f8890]*/
+{
+    if (unlockpt(fd) == -1)
+        return posix_error();
+
+    Py_RETURN_NONE;
+}
+#endif /* HAVE_UNLOCKPT */
+
+#ifdef HAVE_PTSNAME
+/*[clinic input]
+os.ptsname
+
+    fd: fildes
+        File descriptor of a master pseudo-terminal device.
+    /
+
+Return the name of the slave pseudo-terminal device.
+
+Performs a ptsname() system call.
+[clinic start generated code]*/
+
+static PyObject *
+os_ptsname_impl(PyObject *module, int fd)
```

Comment:
Note that

> import os
master, slave = os.openpty()
print(os.ttyname(slave))

and

> import os
master, slave = os.openpty()
print(os.ptsname(master))

will give same output, but `os.ptsname()` works when a file descriptor of the slave is not available!

Reply:
**Notes on thread safety**

Should we add something in the docs about thread safety of these functions?

---

The POSIX manpage (https://pubs.opengroup.org/onlinepubs/9699919799/) of `ptsname()` says the following:

> The ptsname() function need not be thread-safe.

FreeBSD's libc and glibc include a reentrant version called `ptsname_r()` corresponding to the pair `ttyname()`, `ttyname_r()`. However, the `ttyname*` pair is POSIX, whereas only `ptsname()` is POSIX. As a result, `ptsname_r()` has portability issues: https://www.gnu.org/software/gnulib/manual/html_node/ptsname_005fr.html. However, the Linux manpage (https://man7.org/linux/man-pages/man3/ptsname.3.html) claims the following:

> ptsname_r() is a Linux extension, that is proposed for inclusion in the next major revision of POSIX.1 (Issue 8).

If `ptsname_r()` becomes standardized, then we can replace our `ptsname()` call in `os.ptsname()` with `ptsname_r()` in the future. Note that the `os.ttyname()` implementation uses `ttyname_r()` instead of `ttyname()`.

---

The POSIX manpage of `ptsname()` also says the following:

> The application shall not modify the string returned. The returned pointer might be invalidated or the string content might be overwritten by a subsequent call to ptsname(). The returned pointer and the string content might also be invalidated if the calling thread is terminated.

In fact, the Linux manpage says the following:

> On success, ptsname() returns a pointer to a string in static storage which will be overwritten by subsequent calls.  This pointer must not be freed.  On failure, NULL is returned.

Therefore, (almost) simultaneous calls to `os.ptsname()` from different threads A and B such that B calls the system `ptsname()` right in between the `ptsname()` and `PyUnicode_DecodeFSDefault()` calls in A (below) is a problematic situation.